Sources/NIOCore/AsyncChannel/AsyncChannelInboundStreamChannelHandler.swift

```diff
@@ -67,10 +67,59 @@ internal final class NIOAsyncChannelInboundStreamChannelHandler<InboundIn: Senda
     @usableFromInline
     let closeRatchet: CloseRatchet
 
+    /// A type indication what kind of transformation to apply to reads.
+    @usableFromInline
+    enum Transformation {
+        /// A synchronous transformation is applied to incoming reads. This is used when bootstrapping
+        case sync((InboundIn) throws -> ProducerElement)
+        /// In the case of protocol negotiation we are applying a future based transformation where we wait for the transformation
+        /// to finish before we yield it to the source.
+        case protocolNegotiation((InboundIn) -> EventLoopFuture<ProducerElement>)
+    }
+
+    /// The transformation applied to incoming reads.
+    @usableFromInline
+    let transformation: Transformation
+
+    @inlinable
+    init(
+        eventLoop: EventLoop,
+        closeRatchet: CloseRatchet,
+        transformationClosure: @escaping (InboundIn) throws -> ProducerElement
+    ) {
+        self.eventLoop = eventLoop
+        self.closeRatchet = closeRatchet
+        self.transformation = .sync(transformationClosure)
+    }
+
     @inlinable
-    init(eventLoop: EventLoop, closeRatchet: CloseRatchet) {
+    init(
+        eventLoop: EventLoop,
+        closeRatchet: CloseRatchet,
+        protocolNegotiationClosure: @escaping (InboundIn) -> EventLoopFuture<ProducerElement>
+    ) where InboundIn == Channel {
         self.eventLoop = eventLoop
         self.closeRatchet = closeRatchet
+        self.transformation = .protocolNegotiation { channel in
+            return protocolNegotiationClosure(channel)
+                .flatMapErrorThrowing { error in
```

Comment:
I think we need to hop to the channel event loop here as the `protocolNegotiationClosure` could return us a future on any loop (and `pipeline.fireErrorCaught` asserts we're on the right EL).